### PR TITLE
Upgrade to newest versions of `jax` + `jaxlib` and add Windows support for JAX Solver

### LINKED ISSUE #3443: Bump `jax` and `jaxlib` versions
Currently, we have pinned `jaxlib==0.4.8` and `jax=0.4.7` in the extras dependencies, but we could look into bumping the versions for a multitude of reasons:

1. CPU-only [native support for amd64 Windows](https://pypi.org/project/jaxlib/#files) is available now which will enable the use of the Jax solver on Windows (CPU-only) with Python 3.9–3.12. We can mark the dependency with a conditional in setup.py as is standard practice so that there are no conflicts based on platforms and Python versions, and continue to support a lower version for Python 3.8 until we drop support for it (#3390). More details below
2. Availability of arm64/aarch64 wheels for Linux has been there since some minor versions already, this will be helpful for the current proposed solution for multi-platform Docker images in #3430 and will make Jax importable in Docker on M-series macOS machines, the current versions fail due to the use of emulation which does not support AVX instructions.
3. GPU support with Jax: this is available by compiling CUDA and CuDNN on WSL2 and installing the GPU-specific wheels from the Google indices or otherwise directly installing CUDA from the wheels. However, WSL2 requires administrator permissions to install on Windows machines. On Windows, there are some community-enabled wheels for `jax` with CUDA 11.8 and 12.1 with CuDNN 8.6 and 8.9 on https://whls.blob.core.windows.net/unstable/index.html and the instructions for their usage are at https://github.com/cloudhan/jax-windows-builder. On M-series macOS, GPU support is available through the Metal plugin https://developer.apple.com/metal/jax/ which was also noted in https://github.com/pybamm-team/PyBaMM/issues/3274#issuecomment-1737005840. Note: both of these are experimental at this time and potentially unstable, but can be documented as such in the installation instructions. The good things are that both of them are in sync and require the same version of `jaxlib`, i.e., `0.4.11`, and that both of them have now been tested and all GPU-related functionalities in the unit tests seem to be working as intended.

(For more details, please see the extended discussion in the `#infrastructure` channel on Slack)

Therefore, the minimum version to update to could be `jaxlib=0.4.11` and `jax==0.4.11` with the internal version requirement being relaxed a bit in #3121. However, it is also easy to target multiple versions with [conditional dependencies](https://hynek.me/articles/conditional-python-dependencies/) so as to accommodate most if not all of these requirements. Jax itself does this and this is a reference [setup.py file, for example](https://github.com/google/jax/blob/9ae5a43341a3d081ebfed68e0c35dbd3853a86d3/setup.py#L88). Considering that GPU support should be kept standalone, it can be given its own extra dependency set, something like `[jax-gpu]`, or just `[gpu]`, considering that only the Jax solver has it.

A configuration in the `setup.py` file could look like this:

```python
extras_require={
        # provides arm macOS wheels for Docker and native CPU-only support on Windows
        "jax": [
            "jax==0.4.18", python_version>'3.9','<3.12'
            "jaxlib==0.4.18",  python_version>'3.9','<3.12'
            "jax==0.4.1X" # some other minor version that includes support for
            #  Python 3.8 so that our installation does not break
        ],
      # provides GPU support for macOS
        "jax-gpu": [
            "jax==0.4.10",
            "jaxlib==0.4.11",
            "jax-metal==0.0.3"
        ],
```

Otherwise, given the tricky requirements for pre-compiling CUDA and CuDNN, a `nox` session or a helper script can be provided that can be used to fetch the correct versions with `pip` based on the platform.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

Closes #3443. With this PR, I have updated the versions for `jax` and `jaxlib` to match the latest version(s) including support for Windows and for retaining support on Python `3.8`, which we still support.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
It looks like I misread the `jaxlib` release notes since they do not appear to have Python 3.8 support on Windows; it is for Python 3.9+. The JAX developers often delete the releases from PyPI due to storage constraints and provide an alternate index here: https://storage.googleapis.com/jax-releases/jax_releases.html, browsing through which confirmed my suspicion. 

The way around this is to either drop Python 3.8 support (big breaking change) or provide Windows Jax support for just Python 3.9–3.11 (more reasonable change) similar to how `[odes]` will be available for Python <3.12 in #3531. @Saransh-cpp and @brosaplanella: your thoughts will be appreciated on whether this is a good idea, following which I can make the necessary changes.

As I had mentioned, I did start working on rewriting the installation guide side-by-side; however, it shall shift almost all of the pages and their links and would be better as a separate PR. It may be merged in a follow-up queue shortly after the completion of this PR. 

<hr>

P.S. Stepping a bit aside from relevant discussion pertaining to the changes here, but @Saransh-cpp previously mentioned https://scientific-python.org/specs/spec-0000/, which suggested dropping Python 3.9 support by the end of the year since its support window has now ended. While we may not want to do that yet, dropping Python 3.8 could be a reasonable idea since 1. all our core dependencies have done so now, `casadi` was the last out of the pack, and 2. it is generally good practice for packages in the scientific Python ecosystem to be on track with each other and support an array of not more than four major Python versions at a time. It would be nice to discuss this further in the December monthly meeting.

File: setup.py
Comment (agriyakhetarpal) [reply setup.py:269]: I would suggest keeping the pin. Keeping no bounds at all would break things on a future 0.5.X release of `jaxlib`, plus the current GPU support with the Windows community wheels and the macOS Metal plugin is still going to lag behind the current 0.4 releases, so we would want to stay within >=0.4, <0.5 borders so as to not promote different API behaviour for any Jax-related functions.

Comment (kratman) [reply setup.py]: I was not suggesting that we don't pin. Rather we don't need a comment explaining that pinning avoids breaking the build. I like reducing comments whenever possible because in most projects people forget to update them.

**Comment by kratman** [comment]
My 2 cents on dropping 3.8/3.9 is that any change like this will break python environments for some people. Probably better to do both at once than drop one, wait then drop the other. Probably less painful to get it done all at once for users with old versions.

**Comment by agriyakhetarpal** [comment]
> My 2 cents on dropping 3.8/3.9 is that any change like this will break python environments for some people. Probably better to do both at once than drop one, wait then drop the other. Probably less painful to get it done all at once for users with old versions.

Ah I see, are you suggesting we should drop both 3.8 and 3.9? I don't think that would be a good idea since not all of our dependencies will be ready for Python 3.12. That leaves developers and users with less leeway to work between different Python versions, since supporting just 3.10 and 3.11 would be too little for a Python package like PyBaMM that is also useful as a library. 

Python 3.8 support windows had closed a long time back and its EOL is less than a year away (October 2024) which is usually the recommended time for dropping support, but the Scientific Python ecosystem moves quicker and has its own, steadfast ways of proceeding with such things. While Python 3.9 also has a closed support window now, we have two years to reach EOL for it so the day to remove it from PyBaMM is quite far (not until Q3 2025 as the farthest estimate) if we decide to not follow the aforementioned SPEC 0000 in the releases starting next year.

**Comment by Saransh-cpp** [comment]
> The way around this is to either drop Python 3.8 support (big breaking change) or provide Windows Jax support for just Python 3.9–3.11 (more reasonable change) similar to how [odes] will be available for Python <3.12 in https://github.com/pybamm-team/PyBaMM/pull/3531. @Saransh-cpp and @brosaplanella: your thoughts will be appreciated on whether this is a good idea, following which I can make the necessary changes.

Supporting Jax on Python 3.9+ for Windows sounds like a better option.

> Stepping a bit aside from relevant discussion pertaining to the changes here, but @Saransh-cpp previously mentioned https://scientific-python.org/specs/spec-0000/, which suggested dropping Python 3.9 support by the end of the year since its support window has now ended. While we may not want to do that yet, dropping Python 3.8 could be a reasonable idea since 1. all our core dependencies have done so now with casadi being the last out of the pack, and 2. it is generally good practice for packages in the scientific Python ecosystem to be on track with each other and support an array of not more than four major Python versions at a time. It would be nice to discuss this further in the December monthly meeting.

I'm really not a big fan of the Python version's drop schedule listed in SPEC 0000. Most of the libraries I am involved with still support Python 3.8 and will continue until it reaches EOL. I'll ask around and see what the maintainers of other ecosystems think about this, but until then, let's keep supporting Python 3.8.

**Comment by agriyakhetarpal** [comment]
I have added support for `jax==0.4.20`+ `jaxlib==0.4.20` for Python 3.9+ and the PR should be ready for review, but I would suggest holding off until merging this, because I am working on a branch locally checked out from this branch, where I am updating the installation guide with a lot of changes. When I open a PR for that, I can keep both of the branches synced up together so that they can be merged one after the other without conflicts.

**Comment by agriyakhetarpal** [comment]
Discussion from the developer meeting today: there are some changes in the JAX API between `0.4.14` and `0.4.16` plus progress on #2282 is being tested on `jaxlib>=0.4.16` onwards, therefore those changes can be deemed breaking with the version being bumped here to `0.4.20`

This means we will not be able to support `[jax]` on Python 3.8 as mentioned above (we can for Python 3.9–3.11, I will make a note of this in the upcoming revisions for the installation guide). The previously mentioned Apple Metal / Windows unofficial GPU wheels won't get support either. GPU support on WSL and with TPUs is unaffected, of course. For GPU support on experimental platforms, all we can hope for is that the Windows community wheels get bumped (latest version: `0.4.11`) and the Metal support continues to mature in the coming years (`0.4.11` is the latest version there too).

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @agriyakhetarpal! Looks good!

**Comment by agriyakhetarpal** [comment]
It's not blocked by anyone but me. I have to update the installation guide and restructure it to take into account both this PR and #3531. We can merge this after that – though I will still request some more reviews here as I had mentioned in the meeting

**Review by jsbrittain — APPROVED** [review]
@agriyakhetarpal These changes look good to me, thanks.

**Review by BradyPlanden — APPROVED** [review]
Nice one, LGTM!

**Pull request merged**
